experimental/sev_guest/src/crypto.rs

```diff
@@ -71,61 +68,70 @@ impl GuestMessageEncryptor {
         })
     }
 
-    /// Creates an encrypted `GuestMessage` from the provided message.
+    /// Creates an encrypted payload from the provided message and writes that to the target's
+    /// payload field. It also upates the target's header with the appropriate information:
+    /// message type, message size and sequence number.
     ///
     /// The sequence number is incremented automatically if the operation is successful.
-    pub fn encrypt_message<M: AsBytes + Message>(
+    ///
+    /// We consume the input message because we encrypt its memory in place before copying it to the
+    /// payload buffer that is shared with the hypervisor.
+    pub fn encrypt_message<M: AsBytes + FromBytes + Message>(
         &mut self,
-        message: &M,
-    ) -> Result<GuestMessage, &'static str> {
-        let mut result = GuestMessage::new();
-        let bytes = message.as_bytes();
-        result.header.auth_header.message_type = M::get_message_type() as u8;
-        let message_size = bytes.len();
-        result.header.auth_header.message_size = message_size as u16;
-        result.header.sequence_number = self.sequence_number + 1;
-        let mut iv_bytes = [0u8; <Aes256 as BlockSizeUser>::BlockSize::USIZE];
-        iv_bytes[0..size_of::<u64>()].copy_from_slice(result.header.sequence_number.as_bytes());
+        mut message: M,
+        destination: &mut GuestMessage,
+    ) -> Result<(), &'static str> {
+        let buffer = message.as_bytes_mut();
+        destination.header.auth_header.message_type = M::get_message_type() as u8;
+        let message_size = buffer.len();
+        destination.header.auth_header.message_size = message_size as u16;
+        destination.header.sequence_number = self.sequence_number + 1;
+        let mut iv_bytes = [0u8; IV_SIZE];
+        iv_bytes[0..size_of::<u64>()]
+            .copy_from_slice(destination.header.sequence_number.as_bytes());
         let nonce = Nonce::from_slice(&iv_bytes[..]);
-        let associated_data = result.header.auth_header.as_bytes();
-        result.payload[0..message_size].copy_from_slice(bytes);
-        let buffer = &mut result.payload[0..message_size];
+        let associated_data = destination.header.auth_header.as_bytes();
         let auth_tag = self
             .cipher
             .encrypt_in_place_detached(nonce, associated_data, buffer)
             .map_err(|_| "Message encryption failed")?;
```

Comment:
not this PR, but should this error be printed too? or may it contain sensitive data?

Reply:
This error type is deliberately opaque, so printing it will not be useful.  See https://docs.rs/aead/latest/aead/struct.Error.html

Reply:
It would be nice to show that by pattern matching on it with the struct type, which shows that there are no other fields in that error.

Reply:
Just to clarify: do you mean just changing it to `.map_err(|_: aead::Error| "Message encryption failed")?;`, or something else?

Reply:
`map_err(|aead::Error| "Message encryption failed")?` would this work?